sapi/fpm/fpm/fpm_unix.c

```diff
@@ -392,6 +402,93 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
 }
 /* }}} */
 
+#if HAVE_FPM_CPUAFFINITY
+static long fpm_cpumax(void)
+{
+	static long cpuid = LONG_MIN;
+	if (cpuid == LONG_MIN) {
+		cpu_set_t cset;
+#if defined(HAVE_SCHED_SETAFFINITY)
+		if (sched_getaffinity(0, sizeof(cset), &cset) == 0) {
+#elif defined(HAVE_CPUSET_SETAFFINITY)
+		if (cpuset_getaffinity(CPU_LEVEL_WHICH, CPU_WHICH_PID, -1, sizeof(cset), &cset) == 0) {
+#endif
+			cpuid = CPU_COUNT(&cset);
+		} else {
+			cpuid = -1;
+		}
+	}
+
+	return cpuid;
+}
+
+static void fpm_cpuaffinity_init(cpu_set_t *c)
+{
+	CPU_ZERO(c);
+}
+
+static void fpm_cpuaffinity_add(cpu_set_t *c, int min, int max)
+{
+	int i;
+
+	for (i = min; i <= max; i ++) {
+		if (!CPU_ISSET(i, c)) {
+			CPU_SET(i, c);
+		}
+	}
+}
+
+static int fpm_cpuaffinity_set(cpu_set_t *c)
+{
+#if defined(HAVE_SCHED_SETAFFINITY)
+	return sched_setaffinity(0, sizeof(c), c);
+#elif defined(HAVE_CPUSET_SETAFFINITY)
+	return cpuset_setaffinity(CPU_LEVEL_WHICH, CPU_WHICH_PID, -1, sizeof(c), c);
+#endif
+}
+
+static int fpm_setcpuaffinity(char *cpu_list)
+{
+	char *token, *buf;
+    cpu_set_t c;
+	int r, cpumax, min, max;
+
+	r = -1;
+	cpumax = fpm_cpumax();
+
+	fpm_cpuaffinity_init(&c);
+	token = php_strtok_r(cpu_list, ",", &buf);
+
+	while (token) {
+		char *cpu_listsep;
+
+		min = strtol(token, &cpu_listsep, 0);
```

Comment:
this is too forgiving and allows latters in the string. For example `abbbbb123,1`  results in min and max 0 for the first element and 1 for min max second element. Obviously this should be error. I converted this to a small program:
```c
#include <errno.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>

char *php_strtok_r(char *s, const char *delim, char **last)
{
    char *spanp;
    int c, sc;
    char *tok;

    if (s == NULL && (s = *last) == NULL) {
		return NULL;
    }

    /*
     * Skip (span) leading delimiters (s += strspn(s, delim), sort of).
     */
cont:
    c = *s++;
    for (spanp = (char *)delim; (sc = *spanp++) != 0; ) {
		if (c == sc) {
			goto cont;
		}
    }

    if (c == 0)		/* no non-delimiter characters */ {
		*last = NULL;
		return NULL;
    }
    tok = s - 1;

    /*
     * Scan token (scan for delimiters: s += strcspn(s, delim), sort of).
     * Note that delim must have one NUL; we stop if we see that, too.
     */
    for (;;)
    {
		c = *s++;
		spanp = (char *)delim;
		do {
			if ((sc = *spanp++) == c) {
				if (c == 0) {
					s = NULL;
				} else {
					char *w = s - 1;
					*w = '\0';
				}
				*last = s;
				return tok;
			}
		}
		while (sc != 0);
    }
    /* NOTREACHED */
}

int parse(const char *cpu_list_const, int cpumax)
{
    char *token, *buf;
	int min, max;

	int rc = 0;

    char *cpu_list = strdup(cpu_list_const);

    printf("\nparsing %s\n", cpu_list);

	token = php_strtok_r(cpu_list, ",", &buf);

	while (token) {
		char *cpu_listsep;

		min = strtol(token, &cpu_listsep, 0);
		if (errno || min < 0 || min > cpumax) {
			rc = -1;
		}
		max = min;
		if (*cpu_listsep == '-') {
			if (strlen(cpu_listsep) > 1) {
				char *err;
				max = strtol(cpu_listsep + 1, &err, 0);
				if (errno || *err != '\0' || max < min || max > cpumax) {
                    rc = -1;
                    goto end;
				}
			} else {
				rc = -1;
                goto end;
			}
		}

		printf("min: %d, max: %d\n", min, max);

		token = php_strtok_r(NULL, ",", &buf);

    }

end:
    printf("result: %d, cpu list: %s\n", rc, cpu_list);
    free(cpu_list);

	return rc;

}

int main()
{
    parse("1,2,5", 10);
    parse("1-2", 4);
    parse("1-2,4-5,7-10", 12);
    // error cases
    parse("1-8", 4);
    parse("abbbbb123,1", 10);
    parse("123afafkalf,1ffff", 10);
}
```

which results to 
```
parsing 1,2,5
min: 1, max: 1
min: 2, max: 2
min: 5, max: 5
result: 0, cpu list: 1

parsing 1-2
min: 1, max: 2
result: 0, cpu list: 1-2

parsing 1-2,4-5,7-10
min: 1, max: 2
min: 4, max: 5
min: 7, max: 10
result: 0, cpu list: 1-2

parsing 1-8
result: -1, cpu list: 1-8

parsing abbbbb123,1
min: 0, max: 0
min: 1, max: 1
result: 0, cpu list: abbbbb123

parsing 123afafkalf,1ffff
min: 123, max: 123
min: 1, max: 1
result: -1, cpu list: 123afafkalf
```

This needs to be fixed.